File: js/의상(해시).js
// 1.	초기값 obj = {} (빈 객체).
// 2.	첫 번째 아이템 ["hat", "headgear"]:
// •	t[1] = "headgear".
// •	obj["headgear"]가 없으므로 1로 초기화.
// •	결과: { "headgear": 1 }.
// 3.	두 번째 아이템 ["sunglasses", "eyewear"]:
// •	t[1] = "eyewear".
// •	obj["eyewear"]가 없으므로 1로 초기화.
// •	결과: { "headgear": 1, "eyewear": 1 }.
// 4.	세 번째 아이템 ["turban", "headgear"]:
// •	t[1] = "headgear".
// •	obj["headgear"]가 이미 있으므로 +1.
// •	결과: { "headgear": 2, "eyewear": 1 }.


function solution(clothes) {
    return Object.values(clothes.reduce((obj, t) => {
      obj[t[1]] = obj[t[1]] ? obj[t[1]] + 1 : 1;
      return obj;
    }, {})).reduce((a,b) => a * (b+1), 1)-1;
}

File: python/의상(해시).py
# Counter : 데이터의 개수를 셀 때 유용
# Counter(['red', 'blue', 'red', 'green', 'blue', 'blue'])
# -> Counter({'blue': 3, 'red': 2, 'green': 1})

# reduce : 여러 개의 데이터를 대상으로 주로 누적 집계에 사용
# reduce(집계 함수, 순회 가능한 데이터[, 초기값])
# 집계 함수는 두개의 인자를 받음
# 첫번째 인자 : 누적자(accumulator) - 함수 실행의 시작부터 끝까지 계속 재사용 값
# 두번째 인자 : 현재값(current value) - 현재값은 루프 돌면서 계속해서 바뀌는 값
# -> reduce(lambda acc, cur: acc + cur["age"], users, 0)

from collections import Counter
from functools import reduce

def solution(clothes):
    cnt = Counter([kind for name, kind in clothes])
    answer = reduce(lambda x,y: x*(y+1), cnt.values(), 1) -1
    return answer